File: Basic/src/main/java/Generic2.java
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;

/**
 * Created by tony on 2019-06-23.
 */
public class Generic2 {

    public static void main(String[] args) {

        GenericChild child = new GenericChild();
        child.printType();
    }
}

class Father<T> {}

class GenericChild extends Father<String>{

    public void printType() {

        Type genType = getClass().getGenericSuperclass();
        System.out.println(genType);
        Type params = ((ParameterizedType) genType).getActualTypeArguments()[0];
        System.out.println(params);
    }
}

File: Basic/src/main/kotlin/generic/Generic2.kt
package generic

import java.lang.reflect.ParameterizedType

/**
 * Created by tony on 2019-06-23.
 */
object Generic2 {

    @JvmStatic
    fun main(args: Array<String>) {

        val child = GenericChild()
        child.printType()
    }
}

open class Father<T>

class GenericChild : Father<String>() {

    fun printType() {

        val genType = javaClass.genericSuperclass
        println(genType)
        val params = (genType as ParameterizedType).actualTypeArguments[0]
        println(params)
    }
}